bindings/profilers/wall.cc

```diff
@@ -25,213 +26,455 @@
+    // that can belong to this sample. Use half the distance to neighboring
+    // samples, or barring that half the sampling interval
+    uint64_t earliest =
+        1000 * (i > 0 ? (sampleTimestamp + (profile->GetSampleTimestamp(i - 1) -
+                                            zeroSampleTime)) /
+                            2
+                      : (sampleTimestamp - halfInterval));
+    uint64_t latest =
+        1000 * (i < sampleCount - 1
+                    ? (sampleTimestamp +
+                       (profile->GetSampleTimestamp(i + 1) - zeroSampleTime)) /
+                          2
+                    : (sampleTimestamp + halfInterval));
+
+    // This loop will drop all contexts that are too old to be associated with
+    // the current sample; associate those (normally one) that are close enough
+    // to it in time, and stop as soon as it sees a context that's too recent
+    // for this sample.
+    for (;;) {
+      auto contextTimestamp = sampleContext.timestamp - zeroContextTime;
+      if (contextTimestamp < earliest) {
+        // Current sample context is too old (closer to the previous sample than
+        // to this one), discard it and fetch the next one.
+        if (contexts.empty()) {
+          return labelSetsByNode;
+        }
+        sampleContext = contexts.pop_front();
+      } else if (contextTimestamp >= latest) {
+        // Current sample context is too recent (closer to the next sample than
+        // to this one), we'll try to match it to the next sample.
+        break;
+      } else {
+        // This sample context is the closest to this sample.
+        if (sampleContext.labels) {
+          auto it = labelSetsByNode.find(sample);
+          Local<Array> array;
+          if (it == labelSetsByNode.end()) {
+            array = Nan::New<Array>();
+            labelSetsByNode.emplace(sample, array);
+          } else {
+            array = it->second;
+          }
+          Nan::Set(
+              array, array->Length(), sampleContext.labels.get()->Get(isolate));
+        }
+        // Sample context was consumed, fetch the next one
+        if (contexts.empty()) {
+          return labelSetsByNode;
+        }
+        sampleContext = contexts.pop_front();
+        break;  // don't match more than one context to one sample
+      }
+    }
+  }
+  // Push the last popped sample context back into the ring to be used by the
+  // next profile
+  contexts.push_front(std::move(sampleContext));
+  return labelSetsByNode;
 }
 
-WallProfiler::WallProfiler(int interval) : samplingInterval(interval) {}
+WallProfiler::WallProfiler(int intervalMicros, int durationMicros)
+    : samplingInterval(intervalMicros),
+      contexts(durationMicros * 2 / intervalMicros) {}
 
 WallProfiler::~WallProfiler() {
-  Dispose();
+  Dispose(nullptr);
 }
 
-void WallProfiler::Dispose() {
+template <typename F>
+void updateProfilers(F updateFn) {
+  auto currProfilers = profilers.load();
```

Comment:
I think this line needs to be moved inside the CAS loop, otherwise it might loop indefinitely.